File: full-vineflower/me/ag2s/epublib/util/commons/io/ByteOrderMark.java
package me.ag2s.epublib.util.commons.io;

import java.io.Serializable;
import java.util.Locale;

public class ByteOrderMark implements Serializable {
   private static final long serialVersionUID = 1L;
   public static final ByteOrderMark UTF_8 = new ByteOrderMark("UTF-8", 239, 187, 191);
   public static final ByteOrderMark UTF_16BE = new ByteOrderMark("UTF-16BE", 254, 255);
   public static final ByteOrderMark UTF_16LE = new ByteOrderMark("UTF-16LE", 255, 254);
   public static final ByteOrderMark UTF_32BE = new ByteOrderMark("UTF-32BE", 0, 0, 254, 255);
   public static final ByteOrderMark UTF_32LE = new ByteOrderMark("UTF-32LE", 255, 254, 0, 0);
   public static final char UTF_BOM = '\ufeff';
   private final String charsetName;
   private final int[] bytes;

   public ByteOrderMark(final String charsetName, final int... bytes) {
      if (charsetName == null || charsetName.isEmpty()) {
         throw new IllegalArgumentException("No charsetName specified");
      } else if (bytes != null && bytes.length != 0) {
         this.charsetName = charsetName;
         this.bytes = new int[bytes.length];
         System.arraycopy(bytes, 0, this.bytes, 0, bytes.length);
      } else {
         throw new IllegalArgumentException("No bytes specified");
      }
   }

   public String getCharsetName() {
      return this.charsetName;
   }

   public int length() {
      return this.bytes.length;
   }

   public int get(final int pos) {
      return this.bytes[pos];
   }

   public byte[] getBytes() {
      byte[] copy = new byte[this.bytes.length];

      for (int i = 0; i < this.bytes.length; i++) {
         copy[i] = (byte)this.bytes[i];
      }

      return copy;
   }

   @Override
   public boolean equals(final Object obj) {
      if (!(obj instanceof ByteOrderMark)) {
         return false;
      } else {
         ByteOrderMark bom = (ByteOrderMark)obj;
         if (this.bytes.length != bom.length()) {
            return false;
         } else {
            for (int i = 0; i < this.bytes.length; i++) {
               if (this.bytes[i] != bom.get(i)) {
                  return false;
               }
            }

            return true;
         }
      }
   }

   @Override
   public int hashCode() {
      int hashCode = this.getClass().hashCode();

      for (int b : this.bytes) {
         hashCode += b;
      }

      return hashCode;
   }

   @Override
   public String toString() {
      StringBuilder builder = new StringBuilder();
      builder.append(this.getClass().getSimpleName());
      builder.append('[');
      builder.append(this.charsetName);
      builder.append(": ");

      for (int i = 0; i < this.bytes.length; i++) {
         if (i > 0) {
            builder.append(",");
         }

         builder.append("0x");
         builder.append(Integer.toHexString(0xFF & this.bytes[i]).toUpperCase(Locale.ROOT));
      }

      builder.append(']');
      return builder.toString();
   }
}

File: src/main/kotlin/me/ag2s/epublib/util/commons/io/ByteOrderMark.kt
package me.ag2s.epublib.util.commons.io

import java.io.Serializable
import java.util.Locale

class ByteOrderMark(charsetName: String?, vararg bytes: Int) : Serializable {
    private val charsetName: String
    private val bytes: IntArray
    init {
        require(!charsetName.isNullOrEmpty()) { "No charsetName specified" }
        require(bytes.isNotEmpty()) { "No bytes specified" }
        this.charsetName = charsetName
        this.bytes = bytes.copyOf()
    }
    fun getCharsetName(): String = charsetName
    fun length(): Int = bytes.size
    operator fun get(position: Int): Int = bytes[position]
    fun getBytes(): ByteArray = bytes.map { it.toByte() }.toByteArray()
    override fun equals(other: Any?): Boolean = other is ByteOrderMark && bytes.contentEquals(other.bytes)
    override fun hashCode(): Int = javaClass.hashCode() + bytes.sum()
    override fun toString(): String = buildString { append(javaClass.simpleName).append('[').append(charsetName).append(": "); bytes.forEachIndexed { index, value -> if (index > 0) append(','); append("0x").append((value and 0xff).toString(16).uppercase(Locale.ROOT)) }; append(']') }
    companion object {
        private const val serialVersionUID = 1L
        @JvmField val UTF_8 = ByteOrderMark("UTF-8", 239, 187, 191)
        @JvmField val UTF_16BE = ByteOrderMark("UTF-16BE", 254, 255)
        @JvmField val UTF_16LE = ByteOrderMark("UTF-16LE", 255, 254)
        @JvmField val UTF_32BE = ByteOrderMark("UTF-32BE", 0, 0, 254, 255)
        @JvmField val UTF_32LE = ByteOrderMark("UTF-32LE", 255, 254, 0, 0)
        const val UTF_BOM = '\ufeff'
    }
}
